onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_helper.cc

```diff
@@ -0,0 +1,230 @@
+#include <iostream>
+
+namespace onnxruntime {
+
+// The newly-built graph has not yet being resolved by Graph::Resolve(), so we can't leverage ORT Graph IsInputInitializerOrOutput() API.
+// We have to do it by ourselves.
+bool TensorrtExecutionProvider::IsInputInitializerOrOutput(Graph* graph,
+                                                           const std::string& name,
+                                                           bool check_ancestors,
+                                                           std::unordered_map<std::string, std::unique_ptr<SubGraphContext>>& subgraph_context_map) const {
+  return IsLocalValue(graph, name, subgraph_context_map) ||
+         (check_ancestors && graph->ParentGraph() != nullptr &&
+          IsInputInitializerOrOutput(graph->MutableParentGraph(), name, check_ancestors, subgraph_context_map));
+}
+
+// The newly-built graph has not yet being resolved by Graph::Resolve(), so we can't leverage ORT Graph IsOuterScopeValues() API.
+// We have to do it by ourselves.
+bool TensorrtExecutionProvider::IsOuterScopeValue(Graph* graph,
+                                                  const std::string& name,
+                                                  std::unordered_map<std::string, std::unique_ptr<SubGraphContext>>& subgraph_context_map) const {
+  return graph->ParentGraph() != nullptr && IsInputInitializerOrOutput(graph->MutableParentGraph(), name, true, subgraph_context_map);
+}
+
+// The newly-built graph has not yet being resolved by Graph::Resolve(), so we can't leverage ORT Graph IsLocalValue() API.
+// We have to do it by ourselves.
+bool TensorrtExecutionProvider::IsLocalValue(Graph* graph,
+                                             const std::string& name,
+                                             std::unordered_map<std::string, std::unique_ptr<SubGraphContext>>& subgraph_context_map) const {
+  std::string subgraph_name = graph->Name();
+
+  if (subgraph_context_map.find(subgraph_name) == subgraph_context_map.end()) {
+    return false;
+  }
+  SubGraphContext* context = subgraph_context_map.at(subgraph_name).get();
+  return context->output_args.find(name) != context->output_args.cend() ||
+         context->inputs_and_initializers.find(name) != context->inputs_and_initializers.cend();
+}
+
+/**
+ * Set inputs, initializers and outputs for all subgraphs during TensorrtExecutionProvider::GetSupportedList() and save those information in subgraph context data structure.
+ * It's useful for building a valid graph and make Graph::Resolve() happy especially when dealing with nested control-flow op graph.
+ */
+void TensorrtExecutionProvider::BuildSubGraphContext(Graph* graph,
+                                                     std::unordered_map<std::string, std::unique_ptr<SubGraphContext>>& subgraph_context_map) const {
+  // Iterate all the nodes and recurse into inner most subgraph first
+  for (int i = 0; i < graph->MaxNodeIndex(); ++i) {
```

Comment:
Can you iterate graph->Nodes() instead as that already handles skipping invalid nodes?

Reply:
yes, then I need to add a new wrapped type of `GraphNodes` for provider to use `graph->Nodes()`

Reply:
It's fine to do it this way if the provider interface doesn't have that type, although I'm a little surprised it doesn't. Maybe the EP code that uses that interface does things via GraphViewer which filters out invalid nodes IIRC.

Reply:
I can still add GraphNodes and its iterator to the provider interface if I have time. Probably in next PR.